File: Src/DetailedSamples/Visual Basic .NET/Multiple/Xceed Windows Explorer/Utils/TreeView/GZipArchiveTreeViewNode.vb
'
 '* Xceed Zip for .NET - Xceed Windows Explorer sample application
 '* Copyright (c) 2006 - Xceed Software Inc.
 '*
'* [GZipArchiveTreeViewNode.vb]
 '*
 '* Implementation of the AbstractTreeViewNode class and represents a GZipArchive.
 '*
 '* This file is part of Xceed Zip for .NET. The source code in this file
 '* is only intended as a supplement to the documentation, and is provided
 '* "as is", without warranty of any kind, either expressed or implied.
 '


Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Windows.Forms
Imports Xceed.FileSystem
Imports Xceed.Zip
Imports Xceed.Tar
Imports Xceed.FileSystem.Samples.Utils.ListView
Imports Xceed.GZip

Namespace Xceed.FileSystem.Samples.Utils.TreeView
  Public Class GZipArchiveTreeViewNode : Inherits AbstractTreeViewNode
    #Region "CONSTRUCTORS"

    Public Sub New(ByVal file As AbstractFile)
      Me.New(file, Nothing)
    End Sub

    Public Sub New(ByVal file As AbstractFile, ByVal contentListView As System.Windows.Forms.ListView)
      If file Is Nothing Then
        Throw New ArgumentNullException("file")
      End If

      m_item = file
      m_contentListView = contentListView

      Me.RefreshIcon(False)
      Me.RefreshText()

      ' Display a [+] by default
      Me.Nodes.Add(String.Empty)

      ' Initialize the icon updater.
      m_iconUpdater = New TreeViewIconUpdater(Me)
    End Sub

    #End Region ' CONSTRUCTORS

    #Region "PUBLIC PROPERTIES"

    ''' <summary>
    ''' Gets the AbstractFolder this node represent.
    ''' </summary>
    Public Overrides ReadOnly Property Folder() As AbstractFolder
      Get
        Return m_gzipArchive
      End Get
    End Property

    #End Region ' PUBLIC PROPERTIES

    #Region "PUBLIC METHODS"

    Public Overrides Sub InitializeFolder()
      If Not m_gzipArchive Is Nothing Then
        Return
      End If

      Try
        GZipArchive.AllowMultipleFiles = Options.GZipAllowMultipleFiles

        m_gzipArchive = New GZipArchive(CType(IIf(TypeOf m_item Is AbstractFile, m_item, Nothing), AbstractFile))
      Catch e1 As InvalidGZipStructureException
        MessageBox.Show("The GZip archive is either corrupted or not a valid archive.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
      Catch
        MessageBox.Show("An error occured while reading the archive.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
      End Try
    End Sub

    ''' <summary>
    ''' Refresh the current item with this new FileSystemItem. 
    ''' </summary>
    ''' <param name="item">The new FileSystemItem that this item represent.</param>
    Public Overrides Sub Refresh(ByVal item As FileSystemItem)
      MyBase.Refresh(item)

      ' Reset the archive if it was previously loaded.
      If Not m_gzipArchive Is Nothing Then
        m_gzipArchive = Nothing

        Me.InitializeFolder()
      End If
    End Sub

    #End Region ' PUBLIC METHODS

    #Region "PRIVATE FIELDS"

    Private m_gzipArchive As GZipArchive ' = null

    #End Region ' PRIVATE FIELDS
  End Class
End Namespace

File: Src/DetailedSamples/CSharp/Multiple/Xceed Windows Explorer/Utils/TreeView/GZipArchiveTreeViewNode.cs
/*
 * Xceed Zip for .NET - Xceed Windows Explorer sample application
 * Copyright (c) 2006 - Xceed Software Inc.
 * 
 * [GZipArchiveTreeViewNode.cs]
 * 
 * Implementation of the AbstractTreeViewNode class and represents a GZipArchive.
 * 
 * This file is part of Xceed Zip for .NET. The source code in this file 
 * is only intended as a supplement to the documentation, and is provided 
 * "as is", without warranty of any kind, either expressed or implied.
 */

using System;
using System.IO;
using System.Windows.Forms;
using Xceed.FileSystem;
using Xceed.Zip;
using Xceed.Tar;
using Xceed.FileSystem.Samples.Utils.ListView;
using Xceed.GZip;

namespace Xceed.FileSystem.Samples.Utils.TreeView
{
  public class GZipArchiveTreeViewNode : AbstractTreeViewNode
  {
    #region CONSTRUCTORS

    public GZipArchiveTreeViewNode( AbstractFile file )
      : this( file, null )
    {
    }
  
    public GZipArchiveTreeViewNode( AbstractFile file, System.Windows.Forms.ListView contentListView )
    {
      if( file == null )
        throw new ArgumentNullException( "file" );

      m_item = file;
      m_contentListView = contentListView;

      this.RefreshIcon( false );
      this.RefreshText();

      // Display a [+] by default
      this.Nodes.Add( string.Empty );

      // Initialize the icon updater.
      m_iconUpdater = new TreeViewIconUpdater( this );
    }

    #endregion CONSTRUCTORS

    #region PUBLIC PROPERTIES

    /// <summary>
    /// Gets the AbstractFolder this node represent.
    /// </summary>
    public override AbstractFolder Folder
    {
      get
      {
        return m_gzipArchive;
      }
    }

    #endregion PUBLIC PROPERTIES

    #region PUBLIC METHODS

    public override void InitializeFolder()
    {
      if( m_gzipArchive != null )
        return;

      try
      {
        GZipArchive.AllowMultipleFiles = Options.GZipAllowMultipleFiles;

        m_gzipArchive = new GZipArchive( m_item as AbstractFile );
      }
      catch( InvalidGZipStructureException )
      {
        MessageBox.Show( "The GZip archive is either corrupted or not a valid archive.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error );
      }
      catch
      {
        MessageBox.Show( "An error occured while reading the archive.", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error );
      }
    }

    /// <summary>
    /// Refresh the current item with this new FileSystemItem. 
    /// </summary>
    /// <param name="item">The new FileSystemItem that this item represent.</param>
    public override void Refresh( FileSystemItem item )
    {
      base.Refresh( item );

      // Reset the archive if it was previously loaded.
      if( m_gzipArchive != null )
      {
        m_gzipArchive = null;

        this.InitializeFolder();
      }
    }

    #endregion PUBLIC METHODS
  
    #region PRIVATE FIELDS

    private GZipArchive m_gzipArchive; // = null

    #endregion PRIVATE FIELDS
  }
}